sklearn/tree/_criterion.pyx

```diff
@@ -883,62 +781,55 @@ cdef class RegressionCriterion(Criterion):
         """Updated statistics by moving samples[pos:new_pos] to the left."""
 
         cdef DOUBLE_t* y = self.y
-        cdef SIZE_t y_stride = self.y_stride
-        cdef DOUBLE_t* sample_weight = self.sample_weight
-
-        cdef SIZE_t* samples = self.samples
         cdef SIZE_t pos = self.pos
         cdef SIZE_t end = self.end
-
-        cdef SIZE_t n_outputs = self.n_outputs
-        cdef double* sum_left = self.sum_left
-        cdef double* sum_right = self.sum_right
-        cdef double* sum_total = self.sum_total
-
-        cdef SIZE_t i = 0
-        cdef SIZE_t p = 0
-        cdef SIZE_t k = 0
+        cdef SIZE_t i, p, k
         cdef DOUBLE_t w = 1.0
-        cdef DOUBLE_t diff_w = 0.0
+        cdef DOUBLE_t y_ik
 
         # Update statistics up to new_pos
         #
         # Given that
         #           sum_left[x] +  sum_right[x] = sum_total[x]
         # and that sum_total is known, we are going to update
-        # label_count_left from the direction that require the least amount
+        # node_sum_left from the direction that require the least amount
         # of computations, i.e. from pos to new_pos or from end to new_po.
 
         if (new_pos - pos) <= (end - new_pos):
             for p in range(pos, new_pos):
-                i = samples[p]
+                i = self.samples[p]
 
-                if sample_weight != NULL:
-                    w = sample_weight[i]
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
 
-                for k in range(n_outputs):
-                    sum_left[k] += w * y[i * y_stride + k]
+                for k in range(self.n_outputs):
+                    y_ik = y[i * self.y_stride + k]
+                    self.node_sum_left[k] += w * y_ik
+                    self.yw_sq_sum_left += w * y_ik * y_ik
```

Comment:
I don't understand your concern. You can't cache the entire vector, because the ordering will change when you split on a different feature.